File: Baekjoon/2178_미로 탐색/_2178_미로탐색.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;

public class _2178_미로탐색 {
	
	static void maze(int graph[][], int start[], int depth[][]) {
		int dx[]= {-1,1,0,0}; //상하좌우
		int dy[]= {0,0,-1,1};
		
		int visited[][]=new int[graph.length][graph[0].length]; // 방문 여부
		visited[start[0]][start[1]]=1;
		
		Queue<int []> queue=new LinkedList<>();
		queue.add(start);
		
		while(queue.size()!=0) {
			int[] temp=queue.poll();
			for(int i=0; i<4; i++) {
				int x=temp[0]+dx[i];
				int y=temp[1]+dy[i];
				if(x>=0&& x<graph.length && y>=0 && y<graph[0].length) { //미로 범위 안이라면
					if(graph[x][y]==1 && visited[x][y]==0) { // 이동할 수 있으며 아직 방문하지 않았다면
						queue.add(new int[] {x,y});
						visited[x][y]=1;
						depth[x][y]=depth[temp[0]][temp[1]]+1;
					}
				}
			}
		}
	}

	public static void main(String[] args) throws IOException{
		BufferedReader bf=new BufferedReader(new InputStreamReader(System.in));
		
		String str=bf.readLine();
		int n=Integer.parseInt(str.split(" ")[0]);
		int m=Integer.parseInt(str.split(" ")[1]);
		
		int graph[][]=new int[n][m];
		for(int i=0; i<n; i++) {
			str=bf.readLine();
			for(int j=0; j<m; j++) {
				graph[i][j]=Integer.parseInt(str.split("")[j]);
			}
		}
		
		int depth[][]=new int[n][m];
		int start[]= {0,0}; //출발
		
		maze(graph, start, depth);
		
		System.out.println(depth[n-1][m-1]+1);
	}
}

File: Baekjoon/2178_미로 탐색/_2178_미로탐색.py
import sys
from collections import deque

def maze(graph, start, depth):
    dx=[-1,1,0,0] # 상하좌우
    dy=[0,0,-1,1]

    visited = [[0] * len(graph[0]) for i in range(len(graph))] # 방문 여부
    visited[start[0]][start[1]]=1
    queue=deque([start])

    while queue:
        temp=queue.popleft()
        for i in range(4):
            x=temp[0]+dx[i]
            y=temp[1]+dy[i]
            if  x>=0 and x<len(graph) and y>=0 and y<len(graph[0]): # 미로 범위 안이라면
                if graph[x][y]==1 and visited[x][y]==0: # 이동할 수 있으며 아직 방문하지 않았다면
                    queue.append([x,y])
                    visited[x][y]=1
                    depth[x][y]=depth[temp[0]][temp[1]]+1
if __name__=='__main__':
    n,m=map(int, sys.stdin.readline().split())

    graph=[]
    for i in range(n):
        graph.append(list(map(int, sys.stdin.readline().strip())))

    depth = [[0] * m for i in range(n)]
    start=[0,0] # 출발

    maze(graph, start, depth)

    print(depth[n-1][m-1]+1)